### PR TITLE
Windows 10 support

### PR DESCRIPTION
**Pull request by tklengyel**
Using the Rekall profile based method this patch adds support for Windows 10.

./examples/process-list windows10
Process listing for VM windows10 (id=2)
[    4] System (struct addr:ffffe0001469d040)
[  312] smss.exe (struct addr:ffffe0001536e840)
[  384] csrss.exe (struct addr:ffffe00016352080)
[  460] wininit.exe (struct addr:ffffe000163a8080)
[  544] services.exe (struct addr:ffffe000163d9080)
[  560] lsass.exe (struct addr:ffffe000163ff5c0)
[  656] svchost.exe (struct addr:ffffe0001630b840)
[  712] svchost.exe (struct addr:ffffe000168926c0)
[  948] svchost.exe (struct addr:ffffe00016a7f240)
[ 1004] svchost.exe (struct addr:ffffe00016ac3580)
[ 1008] svchost.exe (struct addr:ffffe00016ac1480)
[ 1020] svchost.exe (struct addr:ffffe00016ae5840)
[  540] svchost.exe (struct addr:ffffe00016a70080)
[  480] svchost.exe (struct addr:ffffe0001475e440)
[ 1104] svchost.exe (struct addr:ffffe0001473c080)
[ 1348] spoolsv.exe (struct addr:ffffe0001680b840)
[ 1512] svchost.exe (struct addr:ffffe00016ad5840)
[ 1656] svchost.exe (struct addr:ffffe00016ad3840)
[ 1696] MsMpEng.exe (struct addr:ffffe00016678300)
[ 2220] SearchIndexer. (struct addr:ffffe00017247840)
[ 2856] NisSrv.exe (struct addr:ffffe00017662840)
[ 2432] csrss.exe (struct addr:ffffe000187f4840)
[  388] winlogon.exe (struct addr:ffffe00015fce840)
[ 2536] dwm.exe (struct addr:ffffe00018003840)
[ 2872] LogonUI.exe (struct addr:ffffe00015032080)

./examples/module-list windows10
ntoskrnl.exe
hal.dll
kdcom.dll
mcupdate.dll
werkernel.sys
CLFS.SYS
tm.sys
PSHED.dll
BOOTVID.dll
cmimcext.sys
ntosext.sys
CI.dll
msrpc.sys
FLTMGR.SYS
ksecdd.sys
clipsp.sys
Wdf01000.sys
WDFLDR.SYS
acpiex.sys
WppRecorder.sys
cng.sys
ACPI.sys
WMILIB.SYS
WindowsTrustedRT.sys
WindowsTrustedRTProxy.sys
pcw.sys
msisadrv.sys
pci.sys
isapnp.sys
vdrvroot.sys
ucx01000.sys
pdc.sys
CEA.sys
partmgr.sys
pcmcia.sys
sdbus.sys
vmbus.sys
NDIS.SYS
NETIO.SYS
hvsocket.sys
vmbkmcl.sys
winhv.sys
pciide.sys
PCIIDEX.SYS
spaceport.sys
intelide.sys
volmgr.sys
nvraid.sys
CLASSPNP.SYS
volmgrx.sys
bxvbda.sys
evbda.sys
mountmgr.sys
iaStorV.sys
stornvme.sys
storport.sys
3ware.sys
amdsata.sys
amdxata.sys
amdsbs.sys
arcsas.sys
iaStorAV.sys
lsi_sas.sys
lsi_sas2i.sys
lsi_sas3i.sys
lsi_sss.sys
megasas.sys
megasr.sys
atapi.sys
ataport.SYS
storahci.sys
mvumis.sys
nvstor.sys
percsas2i.sys
percsas3i.sys
SiSRaid2.sys
sisraid4.sys
vsmraid.sys
stexstor.sys
vstxraid.sys
ADP80XX.SYS
HpSAMD.sys
EhStorTcgDrv.sys
EhStorClass.sys
fileinfo.sys
Wof.sys
WdFilter.sys
NTFS.sys
storvsc.sys
usbccgp.sys
USBD.SYS
usbhub.sys
UsbHub3.sys
usbehci.sys
USBPORT.SYS
Fs_Rec.sys
ksecpkg.sys
tcpip.sys
fwpkclnt.sys
wfplwfs.sys
vmstorfl.sys
gagp30kx.sys
agp440.sys
nv_agp.sys
fvevol.sys
uagp35.sys
uliagpkx.sys
volsnap.sys
USBXHCI.SYS
USBSTOR.SYS
uaspstor.sys
storufs.sys
sdstor.sys
sbp2port.sys
rdyboost.sys
mup.sys
intelpep.sys
disk.sys
crashdmp.sys
filecrypt.sys
tbs.sys
Null.SYS
Beep.SYS
BasicDisplay.sys
watchdog.sys
dxgkrnl.sys
BasicRender.sys
Npfs.SYS
Msfs.SYS
tdx.sys
TDI.SYS
netbt.sys
afd.sys
vwififlt.sys
pacer.sys
netbios.sys
rdbss.sys
csc.sys
nsiproxy.sys
npsvctrig.sys
mssmbios.sys
gpuenergydrv.sys
dfsc.sys
ahcache.sys
CompositeBus.sys
kdnic.sys
umbus.sys
i8042prt.sys
mouclass.sys
kbdclass.sys
fdc.sys
usbuhci.sys
intelppm.sys
NdisVirtualBus.sys
swenum.sys
ks.sys
rdpbus.sys
hidusb.sys
HIDCLASS.SYS
HIDPARSE.SYS
mouhid.sys
dump_ataport.sys
dump_atapi.sys
dump_dumpfve.sys
win32k.sys
win32kfull.sys
win32kbase.sys
dxgmms2.sys
TSDDD.dll
luafv.sys
storqosflt.sys
lltdio.sys
mslldp.sys
rspndr.sys
HTTP.sys
WudfPf.sys
bowser.sys
mrxsmb.sys
mrxsmb20.sys
mpsdrv.sys
peauth.sys
srvnet.sys
Ndu.sys
mmcss.sys
srv2.sys
srv.sys
mrxsmb10.sys
tcpipreg.sys
E1G6032E.sys
monitor.sys
tunnel.sys
condrv.sys
WdNisDrv.sys
cdd.dll

### CODE REVIEW

**Comment by tklengyel** [comment]
The kernel loaded into memory seems to be stripped of many of the standard directories normally present. There doesn't seem to be export and debug info present in memory. However, we can still gather the PE GUID which could be used to download a clean copy of the PE executable to gather the PDB GUID from.

./win-guid domid 2
Windows Kernel found @ 0x1043d1000
	Version: 64-bit Windows 10
	PE GUID: 5632d2d17cc000
The debug directory header is not in CodeView format, unable to deal with that!
	Signature: 17744.
	Machine: 34404.
	# of sections: 23.
	# of symbols: 0.
	Timestamp: 1446171345.
	Characteristics: 34.
	Optional header size: 240.
	Optional header type: 0x20b
	Section 1: .text
	Section 2: ERRATA
	Section 3: INITKDBG
	Section 4: POOLCODE
	Section 5: .rdata
	Section 6: .data
	Section 7: .pdata
	Section 8: ALMOSTRO
	Section 9: CACHEALI
	Section 10: PAGELK
	Section 11: PAGE
	Section 12: PAGEKD
	Section 13: PAGEVRFY
	Section 14: PAGEHDLS
	Section 15: PAGEBGFX
	Section 16: PAGEVRFB
	Section 17: .edata
	Section 18: PAGEDATA
	Section 19: PAGEVRFD
	Section 20: INITDATA
	Section 21: INIT
	Section 22: .rsrc
	Section 23: .reloc

**Comment by tklengyel** [comment]
The debug information seem to still be in memory somewhere but not in any of the kernel sections. There is actually multiple hits for the kernel in memory but the debug information is not in any of them. A simple brute-force search quickly reveals it though.

Windows Kernel found @ 0x1043d1000
	Version: 64-bit Windows 10
	PE GUID: 5632d2d17cc000
	PDB GUID: 0de6dc238e194bb78608d54b1e6fa3791
	Kernel filename: ntkrnlmp.pdb
	Multi-processor without PAE
	Signature: 17744.
	Machine: 34404.
	# of sections: 23.
	# of symbols: 0.
	Timestamp: 1446171345.
	Characteristics: 34.
	Optional header size: 240.
	Optional header type: 0x20b
	Section 1: .text
	Section 2: ERRATA
	Section 3: INITKDBG
	Section 4: POOLCODE
	Section 5: .rdata
	Section 6: .data
	Section 7: .pdata
	Section 8: ALMOSTRO
	Section 9: CACHEALI
	Section 10: PAGELK
	Section 11: PAGE
	Section 12: PAGEKD
	Section 13: PAGEVRFY
	Section 14: PAGEHDLS
	Section 15: PAGEBGFX
	Section 16: PAGEVRFB
	Section 17: .edata
	Section 18: PAGEDATA
	Section 19: PAGEVRFD
	Section 20: INITDATA
	Section 21: INIT
	Section 22: .rsrc
	Section 23: .reloc

File: libvmi/os/windows/core.c [hunk 1]
```diff
@@ -50,6 +51,8 @@ win_ver_t ntbuild2version(uint16_t ntbuildnumber)
         case 9200:
         case 9600:
             return VMI_OS_WINDOWS_8;
+        case 18432:
```
Comment (smaresca) [inline libvmi/os/windows/core.c]: Isn't w10 build 10xyx and up? E.g., build 10047?

Comment (tklengyel) [inline libvmi/os/windows/core.c]: I'm getting weird values occasionally, sometimes under 9000. Not sure what's up with that so I just turned this check into a warning.

File: examples/win-guid.c [hunk 2]
```diff
@@ -117,20 +156,48 @@ void print_os_version(vmi_instance_t vmi, addr_t kernel_base_p, uint8_t* pe) {
...
+        uint8_t pe[VMI_PS_4KB];
+        vmi_read_pa(vmi, i, pe, VMI_PS_4KB);
+        uint32_t c = 0;
+        for (c;c<VMI_PS_4KB-12;c++) {
+            if(!strncmp((char*)&pe[c], "ntkrnlmp.pdb", 12) ||
+               !strncmp((char*)&pe[c], "ntoskrnl.pdb", 12) ||
+               !strncmp((char*)&pe[c], "ntkrnlpa.pdb", 12) ||
+               !strncmp((char*)&pe[c], "ntkrpamp.pdb", 12)
+            ) {
+                vmi_read_pa(vmi, i+c - 2*sizeof(uint32_t) - sizeof(struct guid), pdb_header, sizeof(struct cv_info_pdb70)+12);
```
Comment (awsaba) [inline examples/win-guid.c]: Since `pdb_header` can be variable length, the actual size allocated should be passed in as a parameter.

Comment (tklengyel) [inline examples/win-guid.c]: Well, we know what we are looking for is only 12 character long on top of the pdb_header so it's not really variable length.

Comment (awsaba) [inline examples/win-guid.c]: Then at least #define it since it's used in at least 2 places.

Comment (tklengyel) [inline examples/win-guid.c]: Sure but keep in mind this is just an example code.

**Comment by awsaba** [comment]
It would be nice to have a definitive reason for the non-matching NtBuildNumbers, but I still think this is good to merge if it's working with Rekall profiles.

File: libvmi/os/windows/core.c [hunk 3]
```diff
@@ -652,12 +659,14 @@ static status_t
...
-    if (windows->rekall_profile) {
-        return init_from_rekall_profile(vmi);
-    } else {
+    if (windows->rekall_profile)
+        ret = init_from_rekall_profile(vmi);
+
+    /* Fall be here too if the Rekall profile based init fails */
+    if ( VMI_FAILURE == ret )
         return init_from_kdbg(vmi);
-    }
```
Comment (CMF-Marodeur) [inline libvmi/os/windows/core.c]: You are not returning anything if the init_from_rekall_profile succeeds.

File: libvmi/os/windows/core.c (hunk 1 above)
Comment (CMF-Marodeur) [inline libvmi/os/windows/core.c]: Are there any news concerning the weird NtBuildNumbers?
I really would like to see this merged (with all three numbers currently possible for Win10 ) 
> 10240 (TH1)
> 10586 (TH2)
> 14393 (RS1)

In my tests I did not see any other values than the expected ones, so returning VMI_FAILURE if no valid ntbuildnumber is found works well for me.

File: libvmi/os/windows/core.c (hunk 3 above)
Comment (tklengyel) [inline libvmi/os/windows/core.c]: Ops, thx ;)

**Comment by smaresca** [comment]
+2 along with @awsaba . Merging

**Comment by cuiyudaniuHT** [comment]
How to write a win 10 profile? @tklengyel @awsaba  @smaresca

**Pull request merged**